crates/bevy_ecs/src/world/entity_ref.rs

```diff
@@ -2998,6 +3254,187 @@ mod tests {
+        for entity_mut in query.iter_mut(&mut world) {
+            found = true;
+            assert!(entity_mut.get::<TestComponent>().is_none());
+            assert!(matches!(
+                entity_mut.get::<TestComponent2>(),
+                Some(TestComponent2(0))
+            ));
+        }
+
+        assert!(found);
+    }
+
+    // Test that an `EntityMutExcept` that doesn't include a component C among
+    // its exclusions can't coexist with a query for that component.
+    #[test]
+    #[should_panic]
+    fn entity_mut_except_conflicts() {
+        let mut world = World::new();
+        world.init_component::<TestComponent>();
+        world.init_component::<TestComponent2>();
+
+        world.spawn(TestComponent(0)).insert(TestComponent2(0));
+
+        // This should panic, because we have a mutable borrow on
+        // `TestComponent` but have a simultaneous indirect immutable borrow on
+        // that component via `EntityRefExcept`.
+        let system = (
+            QueryParamBuilder::new(|builder| {
+                builder.data::<&mut TestComponent>();
+            }),
+            QueryParamBuilder::new(|builder| {
+                builder.data::<EntityMutExcept<TestComponent2>>();
+            }),
+        )
+            .build_state(&mut world)
+            .build_system(
+                |_: Query<&mut TestComponent>,
+                 mut query: Query<EntityMutExcept<TestComponent2>>| {
+                    for entity_mut in query.iter_mut() {
+                        assert!(entity_mut
+                            .get_mut::<TestComponent2>()
+                            .is_some_and(|component| component.0 == 0));
+                    }
+                },
+            );
+
+        world.run_system_once(system);
+    }
+
+    // Test that an `EntityMutExcept` with an exception for some component C can
+    // coexist with a query for that component C.
+    #[test]
+    fn entity_mut_except_doesnt_conflict() {
+        let mut world = World::new();
+        world.init_component::<TestComponent>();
+        world.init_component::<TestComponent2>();
+
+        world.spawn(TestComponent(0)).insert(TestComponent2(0));
+
+        let system = (
+            QueryParamBuilder::new(|builder| {
+                builder.data::<&mut TestComponent>();
+            }),
+            QueryParamBuilder::new(|builder| {
+                builder.data::<EntityMutExcept<TestComponent>>();
+            }),
+        )
+            .build_state(&mut world)
+            .build_system(
+                |_: Query<&mut TestComponent>, mut query: Query<EntityMutExcept<TestComponent>>| {
+                    for entity_mut in query.iter_mut() {
+                        assert!(entity_mut
+                            .get_mut::<TestComponent2>()
+                            .is_some_and(|component| component.0 == 0));
+                    }
+                },
+            );
+
+        world.run_system_once(system);
+    }
+
```

Comment:
We should have tests that all the get_* methods return none when the component exists and is in the deny list, since this is a safety concern.

Reply:
Added.